### PR TITLE
use only major version for soname

### PR DESCRIPTION
**Pull request by xaiki**
hello NDN team, i sent this mail to ndn-lib last week, but it looks like it never reached there,
please tell me where is the best place to discuss this...

at Endlessm.com we're working on integrating NDN software in our next
release, and part of that works involves pushing software to upstream
Debian. For that we've been working on a few packages.

tl;dr: do we want to commit to bumping soname only when we actually have an
ABI break ?

one of them is the package for ndn-cxx or in debian terms libndn-cxx, the
current ppa packaging calls it just ndn-cxx, but that violates policy (libs
should have the soname in their name).

https://github.com/named-data/ppa-packaging/blob/master/ndn-cxx/debian/control

it also ships binaries in a single pack.

we created a new package for ndn-cxx that is cleaner in respect to Debian
Policy and packaging best practices, it creates an ndn-tools package (maybe
should be called ndn-utils to not conflict with the ndn-tools repo ?) uses a
tool we created: dh-waf to simplify greatly the packaging (and make it
multiarch ready). and finally, that's the purpose of this mail, ships a
patch to set soname to the major version

https://github.com/endlessm/ndn-cxx/blob/T13932-packaging/debian/control
https://github.com/endlessm/dh-waf
https://github.com/endlessm/ndn-cxx/blob/T13932-packaging/debian/patches/make-soname-be-the-major-version

without that patch, every change of version would require a new package
(i.e. libndn-cxx0.5.0 -> libndn-cxx0.5.1 -> libndn-cxx0.5.2...) while i
understand not tracking sonames makes it easier for dev, it makes it much
harder for integration in distros, so i'd like to find a place where it
works for all of us.

how can we find a common ground between not putting burden on dev and on
packaging ? whilst still making it possible to ship code in a distro ?

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

### CODE REVIEW

**Comment by Pesa** [comment]
> one of them is the package for ndn-cxx or in debian terms libndn-cxx, the current ppa packaging calls it just ndn-cxx, but that violates policy

yeah good point, the library pkgs should be called `libndn-cxx` and `libndn-cxx-dev`.

>  it creates an ndn-tools package (maybe should be called ndn-utils to not conflict with the ndn-tools repo ?)

`ndn-tools` is definitely a bad idea, I'd suggest `libndn-cxx-bin` (I've seen other packages do this) or simply `ndnsec`, as it's basically that one tool plus several wrapper scripts.

Please note that the file `ndn-client.conf` *must* be provided by `libndn-cxx`.

> without that patch, every change of version would require a new package
(i.e. libndn-cxx0.5.0 -> libndn-cxx0.5.1 -> libndn-cxx0.5.2...) while i
understand not tracking sonames makes it easier for dev, it makes it much
harder for integration in distros, so i'd like to find a place where it
works for all of us.

I understand your concern very well (I'm a gentoo packager myself). Unfortunately the ndn-cxx project still has an unstable ABI, we break it all the time. I don't think we ever had a release that was binary compatible with the previous one. So effectively the soname *does* change with every release.

**Comment by cawka** [comment]
> hello NDN team, i sent this mail to ndn-lib last week, but it looks like it never reached there,
> please tell me where is the best place to discuss this...

For this specific discussion, I would recommend [nfd-dev mailing list](http://www.lists.cs.ucla.edu/mailman/listinfo/nfd-dev)

> at Endlessm.com we're working on integrating NDN software in our next
> release, and part of that works involves pushing software to upstream
> Debian. For that we've been working on a few packages.
> 
> tl;dr: do we want to commit to bumping soname only when we actually have an
> ABI break ?

ABI is broken upon every minor and patch release.  This is a semi-property of C++ library.  We have investigated this specific issue before and intentionally introduced the full-versioned soname.  This should not be changed.

> one of them is the package for ndn-cxx or in debian terms libndn-cxx, the
> current ppa packaging calls it just ndn-cxx, but that violates policy (libs
> should have the soname in their name).

Thanks for pointing this out.  We should fix the name.  Though, I'm confused about your statement in parenthesis.

> it also ships binaries in a single pack.

What exactly you mean by that?  You mean, the sonamed library and the non-versioned symlink?

> we created a new package for ndn-cxx that is cleaner in respect to Debian
> Policy and packaging best practices, it creates an ndn-tools package (maybe
> should be called ndn-utils to not conflict with the ndn-tools repo ?) uses a
> tool we created: dh-waf to simplify greatly the packaging (and make it
> multiarch ready). and finally, that's the purpose of this mail, ships a
> patch to set soname to the major version

> https://github.com/endlessm/ndn-cxx/blob/T13932-packaging/debian/control

I checked your repo, it is actually called `ndnsec-tools`, so no problem with the name.  I would probably call it just `ndnsec`, but either way is good.

Just in case. If the intention to include debian files in the ndn-cxx repository, then I will be against it. My preference is to keep packaging related, but separated from the actual code, as it is done with ppa-packaging.

> https://github.com/endlessm/dh-waf

I am not too convinced (though not against) having this.  The rules file is not that a big and usually includes several customizations.

> https://github.com/endlessm/ndn-cxx/blob/T13932-packaging/debian/patches/make-soname-be-the-major-version

As I said before, this patch should not be included.

> without that patch, every change of version would require a new package
> (i.e. libndn-cxx0.5.0 -> libndn-cxx0.5.1 -> libndn-cxx0.5.2...) while i
> understand not tracking sonames makes it easier for dev, it makes it much
> harder for integration in distros, so i'd like to find a place where it
> works for all of us.

It is not about taking easier.  The code compiled with 0.5.0 will not work with 0.5.1 or 0.5.2.  At least not at the present moment.  This can change later (after version 1.x), where we will try to ensure ABI compatibility across patch releases, but it is additional complexity.  It is a burden on package management, but it is the required burden.  Anytime the library is updated, the dependencies must be recompiled against the new version or they will be segfaulting.
